src/components/ColorPickerContainer.ts

```diff
@@ -0,0 +1,286 @@
+import { Component, createElement } from "react";
+import { ColorResult } from "react-color";
+
+import { ColorPicker, Mode, PickerType } from "./ColorPicker";
+import { Input } from "./Input";
+import { Button } from "./Button";
+import { Label } from "./Label";
+
+interface WrapperProps {
+    class: string;
+    mxObject?: mendix.lib.MxObject;
+    mxform: mxui.lib.form._FormBase;
+    style: string;
+    readOnly: boolean;
+    friendlyId: string;
+}
+
+interface Nanoflow {
+    nanoflow: object[];
+    paramsSpec: { Progress: string };
+}
+
+export interface ColorPickerContainerProps extends WrapperProps {
+    colorAttribute: string;
+    editable: "default" | "never";
+    format: Format;
+    mode: Mode;
+    type: PickerType;
+    label: string;
+    labelWidth: number;
+    onChangeEvent: onChange;
+    onChangeMicroflow: string;
+    onChangePage: string;
+    onChangeNanoflow: Nanoflow;
+    openPageLocation: "content" | "popup" | "modal";
+    defaultColors: string[];
+}
+
+interface ColorPickerContainerState {
+    color: string;
+    alertMessage?: string;
+    displayColorPicker: boolean;
+}
+
+type Format = "hex" | "rgb" | "rgba";
+type onChange = "doNothing" | "showPage" | "callMicroflow" | "callNanoflow";
+
+export default class ColorPickerContainer extends Component<ColorPickerContainerProps, ColorPickerContainerState> {
+    private subscriptionHandles: number[] = [];
+    private disabled = false;
+    private previousColor = "";
+    private colorChanged = false;
+
+    constructor(props: ColorPickerContainerProps) {
+        super(props);
+
+        this.state = {
+            alertMessage: "",
+            color: "",
+            displayColorPicker: false
+        };
+    }
+
+    render() {
+        const { mxObject, readOnly, colorAttribute, format, type } = this.props;
+        this.disabled = this.props.editable !== "default"
```

Comment:
can extract this line to a function `isReadOnly()` for readability

Reply:
Done